JOSS publication (#1885)
r-lib/lintr: paper/paper.Rmd

---
title: "Static Code Analysis for R"
date: "`r Sys.Date()`"
tags: ["R", "linter", "tidyverse"]

authors:
  - name: Jim Hester
    affiliation: 1
    orcid: ~

affiliations:
  - index: 1
    name: Netflix

output: 
  rticles::joss_article:
    keep_tex: true
bibliography: paper.bib
csl: apa.csl
journal: JOSS
link-citations: yes
---

```{r, warning=FALSE, message=FALSE, echo=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  out.width = "100%",
  dpi = 300,
  comment = "#>"
)

options(width = 60)
```

# Statement of Need

> The R programming language [@base2023] is a popular choice for statistical analysis and visualization, and is used by a wide range of researchers and data scientists. The `{lintr}` package is an open-source R package that provides static code analysis to check for a variety of common problems related to readability, efficiency, consistency, style, etc. It is designed to be easy to use and integrate into existing workflows, and can be run from the command line or used as part of an automated build or continuous integration process. `{lintr}` also integrates with a number of popular IDEs and text editors, such as RStudio and Visual Studio Code, making it convenient for users to run `{lintr}` checks on their code as they work.

Reviewer: if this audience may not be familiar with R, it's important to clarify that R is dynamically-typed, since that colors so much of what we can/can't do with {lintr}.

Reviewer: So far, I don't see any mention that a primary purpose of {lintr} is helping to enforce the tidyverse style guide in particular. I think this is a useful contextual touchstone.

Author: > So far, I don't see any mention that a primary purpose of {lintr} is helping to enforce the tidyverse style guide in particular. I think this is a useful contextual touchstone.

@MichaelChirico It's a bit ironic that (AFAICT) none of the tidyverse packages actually use `{lintr}` in their development workflow 😅

# Features

> There are over 85 linters offered by `{lintr}`!

Reviewer: ```suggestion
As of this writing, `{lintr}` offers 85 linters
```

That's roughly on par with the # I see in e.g. `flake8`, I'm not sure "more than N!" is the right framing :)

```{r}
library(lintr)

length(all_linters())
```

Naturally, we can't discuss all of them here. To see details about all available linters, we encourage readers to see <https://lintr.r-lib.org/dev/reference/index.html#individual-linters>.

We will showcase one linter for each kind of common problem found in R code.

- **Best practices**

`{lintr}` offers linters that can detect problematic antipatterns and suggest alternative patterns that follow best practices.

> For example, usage of vectorized `&` and `|` logical operators in conditional statements is error-prone, and scalar `&&` and `||`, respectively, are to be preferred. The `vector_logic_linter()` linter detects such problematic usages.

Reviewer: If the audience may not know R, it will be good to discuss the difference of vector/scalar operators a bit more in the context of R as a language.

Author: Hmm, I am not sure if we need to get into that. 

Will it be better to use some other linter as an example to avoid having to explain vector/scalar operators?

---
title: "Static Code Analysis for R"
date: "`r Sys.Date()`"
tags: ["R", "linter", "tidyverse"]

> authors:

Author: I have added only a preliminary list of authors. I am not sure about what should either the details (authors can do this themselves) or the order here.

  - name: Jim Hester
    affiliation: 1
    orcid: ~
  - name: Michael Chirico
    affiliation: 2
    orcid: ~
  - name: Alexander Rosenstock
    affiliation: 3
    orcid: ~
  - name: Indrajeet Patil
    orcid: 0000-0003-1995-6531
    affiliation: 4
affiliations:
  - index: 1
    name: Netflix
  - index: 2
    name: Google
  - index: 3
    name: Mathematisches Institut der Heinrich-Heine-Universität Düsseldorf
  - index: 4
    name: Center for Humans and Machines, Max Planck Institute for Human Development, Berlin, Germany
output: 
    md_document:
      variant: "markdown"
      preserve_yaml: true
      standalone: true
bibliography: paper.bib
csl: apa.csl
link-citations: yes
---

```{r, warning=FALSE, message=FALSE, echo=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  out.width = "100%",
  comment = "#>"
)

library(lintr)

withr::local_options(list(
  lintr.format_width = 60L
))
```

# Statement of Need

> R is an interpreted, dynamically typed programming language [@base2023]. It is a popular choice for statistical analysis and visualization, and is used by a wide range of researchers and data scientists. The `{lintr}` package is an open-source R package that provides static code analysis to check for a variety of common problems related to readability, efficiency, consistency, style, etc. In particular, it enforces the [tidyverse style guide]((https://style.tidyverse.org/index.html)). It is designed to be easy to use and integrate into existing workflows, and can be run from the command line or used as part of an automated build or continuous integration process. `{lintr}` also integrates with a number of popular IDEs and text editors, such as RStudio and Visual Studio Code, making it convenient for users to run `{lintr}` checks on their code as they work.

Reviewer: > can be run from the command line

This might make it seem like one could do `./lintr file --linters=assignment_linter()` or something like that, but we don't ship any command-line binary, you have to invoke `Rscript` (right?)

Author: True. I no longer mention the command line here.

# Features

As of this writing, `{lintr}` offers `r length(all_linters())` linters.

```{r}
library(lintr)

length(all_linters())
```

Naturally, we can't discuss all of them here. To see details about all available linters, we encourage readers to see <https://lintr.r-lib.org/dev/reference/index.html#individual-linters>.

We will showcase one linter for each kind of common problem found in R code.

- **Best practices**

> `{lintr}` offers linters that can detect problematic antipatterns and suggest alternative patterns that follow best practices.

Reviewer: ```suggestion
`{lintr}` offers linters that can detect problematic antipatterns and suggest alternatives that follow best practices.
```

For example, expressions like `ifelse(x, TRUE, FALSE)` and `ifelse(x, FALSE, TRUE)` are redundant; just `x` or `!x` suffice in R code where logical vectors are a core data structure. The `redundant_ifelse_linter()` linter detects such problematic usages.

```{r}
lint(
  text = "ifelse(x >= 2.5, TRUE, FALSE)",
  linters = redundant_ifelse_linter()
)
```

> ```{r, eval=FALSE, echo=FALSE}

Author: These code chunks are currently unevaluated because I am not sure if we want to include examples that don't lint. 

If we decide not to, these chunks can be removed.

Reviewer: I lean towards dropping them.

lint(

>   text = "ifelse(x >= 2.5)",

Reviewer: is this meant to have no `yes`/`no` arguments?

  linters = redundant_ifelse_linter()
)
```

- **Efficiency**

> Sometimes the users might not be aware of a more efficient way offered by R for carrying out a computation. `{lintr}` offers linters to provide suggestions to improve code efficiency.

Reviewer: ```suggestion
Sometimes the users might not be aware of a more efficient way offered by R for carrying out a computation. `{lintr}` offers linters to improve code efficiency by avoiding common inefficient patterns.
```

```{r}
lint(
  text = "any(is.na(x), na.rm = TRUE)",
  linters = any_is_na_linter()
)
```

> ```{r, echo=FALSE, eval=FALSE}

Reviewer: Here, the "no lint" example does add value, but maybe better served by explaining in text why `any(is.na())` is deemed inefficient.

lint(
  text = "anyNA(x)",
  linters = any_is_na_linter()
)
```

- **Readability**

> Coders spend significantly more time reading compared to writing code [@mcconnell2004code]. Thus, writing readable code makes the code more maintainable and reduces the possibility of introducing bugs stemming from a poor understanding of the code.

Reviewer: ```suggestion
Coders spend significantly more time reading than writing code [@mcconnell2004code]. Thus, writing readable code makes the code more maintainable and reduces the possibility of introducing bugs stemming from a poor understanding of the code.
```

`{lintr}` provides a number of linters that suggest more readable alternatives. For example, `function_left_parentheses_linter()`.

```{r}

> lint(

Reviewer: I would suggest a more compelling readability linter for the paper, maybe `comparison_negation_linter()`?

  text = "stats::sd (c (x, y, z))",
  linters = function_left_parentheses_linter()
)
```

```{r, echo=FALSE, eval=FALSE}
lint(
  text = "stats::sd(c(x, y, z))",
  linters = function_left_parentheses_linter()
)
```

> - **Tidyverse style**

Reviewer: I think it'd be good to include a "common mistakes" example too. That's a good example of when {lintr} giving early feedback is important (usually these issues would be caught by tests, but (1) that happens "much" later than typing the code (2) assumes high test coverage).

> `{lintr}` also provides linters to enforce the style used throughout the `{tidyverse}` [@Wickham2019] ecosystem of R packages. This style of coding has been outlined in the tidyverse style guide (https://style.tidyverse.org/index.html).

Reviewer: > https://style.tidyverse.org/index.html

Also linked above, should it be a bibliography entry?

Author: Created a bib entry. Currently treating as a manual since it's not a published book.

```{r}
lint(

>   text = "1:3 %>% mean %>% as.character",

Reviewer: Maybe `object_name_linter()` would be a better choice? This one is kind-of a {magrittr} bug, harder to think of it as being a "core" aspect of the tidyverse guide. Whereas tidyverse naming style does "stick out" a bit in R which has naming style wild west.

  linters = pipe_call_linter()
)
```

```{r, echo=FALSE, eval=FALSE}
lint(
  text = "1:3 %>% mean() %>% as.character()",
  linters = pipe_call_linter()
)
```

> # Benefits of using `{lintr}`

Reviewer: One thing I'd also highlight is our design for extensibility -- yes, we enforce the Tidyverse guide by default, but we designed the package to fit many different style guides for any person/team/org that wants to deviate or even to start their own guide.

Author: How about this?

> `{lintr}` is designed for extensibility by allowing users to easily create custom linting rules, enabling flexible and tailored code analysis to suit specific coding conventions, style guides, and project requirements.

---
title: "Static Code Analysis for R"
date: "`r Sys.Date()`"
tags: ["R", "linter", "tidyverse"]
authors:
  - name: Jim Hester
    affiliation: 1
    orcid: ~
  - name: Michael Chirico
    affiliation: 2

>     orcid: ~

Reviewer: ```suggestion
    orcid: 0000-0003-0787-087X
```

bibliography: paper.bib
csl: apa.csl
link-citations: yes
---

```{r, warning=FALSE, message=FALSE, echo=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  out.width = "100%",
  comment = "#>"
)

library(lintr)

withr::local_options(list(
  lintr.format_width = 60L
))
```

# Statement of Need

R is an interpreted, dynamically typed programming language [@base2023]. It is a popular choice for statistical analysis and visualization, and is used by a wide range of researchers and data scientists. The `{lintr}` package is an open-source R package that provides static code analysis to check for a variety of common problems related to readability, efficiency, consistency, style, etc. In particular, it enforces the tidyverse style guide [@Wickham2023]. It is designed to be easy to use and integrate into existing workflows, and can be used as part of an automated build or continuous integration process. `{lintr}` also integrates with a number of popular IDEs and text editors, such as RStudio and Visual Studio Code, making it convenient for users to run `{lintr}` checks on their code as they work.

# Features

As of this writing, `{lintr}` offers `r length(all_linters())` linters.

> ```{r}

Author: It is an overkill, but let's do it anyway, lest it's marked as a lint in future by some doclinter 😉

For example, the style guide recommends using snake_case for identifiers:

```{r}
lint(
  text = "MyVar <- 1L",
  linters = object_name_linter()
)
```

But, as with other linters in this category, the linter can be customized for any other style guide:

```{r}
lint(
  text = "my.var <- 1L",
  linters = object_name_linter(styles = "dotted.case")
)
```

- **Common mistakes**

One category of linters help you detect some common mistakes statically and provide early feedback.

For example, duplicate arguments in function calls can sometimes cause errors:

```{r, error=TRUE}
mean(x = 1:5, x = 2:3)
```

But `duplicate_argument_linter()` can check for this statically:

```{r}
lint(
  text = "mean(x = 1:5, x = 2:3)",
  linters = duplicate_argument_linter()
)
```

Even for cases where there will not be a run-time error, this linter explicitly discourages duplicate arguments.

```{r}
lint(
  text = "list(x = TRUE, x = FALSE)",
  linters = duplicate_argument_linter()
)
```

This is because duplicate-named objects can be hard to work with programmatically and should typically be avoided.

```{r}
l <- list(x = TRUE, x = FALSE)

> l[["x"]]

Author: I agree that this is a better example. I have included it in the paper.

```

# Benefits of using `{lintr}`

There are several benefits to using `{lintr}` to analyze and improve R code. One of the most obvious is that it can help users identify and fix problems in their code, which can save time and effort during the development process. By catching issues early on, `{lintr}` can help prevent bugs and other issues from creeping into code, which can save time and effort when it comes to debugging and testing.

Another benefit of `{lintr}` is that it can help users write more readable and maintainable code. By enforcing a consistent style and highlighting potential issues, `{lintr}` can help users write code that is easier to understand and work with. This is especially important for larger projects or teams, where multiple contributors may be working on the same codebase and it is important to ensure that code is easy to follow and understand.

> `{lintr}` is designed for extensibility by allowing users to easily create custom linting rules, enabling flexible and tailored code analysis to suit specific coding conventions, style guides, and project requirements.

Author: I agree with the first point, but I think the second point/example would be too technical for this type of article (since it involves many concepts that we never discuss here, e.g. XPath) that is supposed to target a wide range of readers. WDYT?

> `{lintr}` is designed for extensibility by allowing users to easily create custom linting rules, enabling flexible and tailored code analysis to suit specific coding conventions, style guides, and project requirements.

Reviewer: Yea, I had a similar worry. How about instead of a specific example, a paragraph at a high enough level about the ability to write custom linters?

> `{lintr}` is designed for extensibility by allowing users to easily create custom linting rules, enabling flexible and tailored code analysis to suit specific coding conventions, style guides, and project requirements.

Author: I have added a new section to discuss this. Can you please have a look?

---
title: "Static Code Analysis for R"
date: "`r Sys.Date()`"
tags: ["R", "linter", "tidyverse"]
authors:
  - name: Jim Hester
    affiliation: 1
    orcid: ~
  - name: Michael Chirico
    affiliation: 2
    orcid: 0000-0003-0787-087X
  - name: Alexander Rosenstock
    affiliation: 3
    orcid: ~
  - name: Indrajeet Patil
    orcid: 0000-0003-1995-6531
    affiliation: 4
affiliations:
  - index: 1
    name: Netflix
  - index: 2
    name: Google
  - index: 3
    name: Mathematisches Institut der Heinrich-Heine-Universität Düsseldorf
  - index: 4
    name: Preisenergie GmbH, Munich, Germany
output: 
    md_document:
      variant: "markdown"
      preserve_yaml: true
      standalone: true
bibliography: paper.bib
csl: apa.csl
link-citations: yes
---

```{r setup, warning=FALSE, message=FALSE, echo=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  out.width = "100%",
  comment = "#>"
)

library(lintr)

withr::local_options(list(
  lintr.format_width = 60L
))
```

# Statement of Need

> R is an interpreted, dynamically-typed programming language [@base2023]. It is a popular choice for statistical analysis and visualization, and is used by a wide range of researchers and data scientists. The `{lintr}` package is an open-source R package that provides static code analysis to check for a variety of common problems related to readability, efficiency, consistency, style, etc. In particular, by default it enforces the tidyverse style guide [@Wickham2023]. It is designed to be easy to use and integrate into existing workflows, and can be used as part of an automated build or continuous integration process. `{lintr}` also integrates with a number of popular IDEs and text editors, such as RStudio and Visual Studio Code, making it convenient for users to run `{lintr}` checks on their code as they work.

Reviewer: Should we add a citation about "static code analysis", e.g. just the Wikipedia page?

Reviewer: Similarly, does it make sense to mention popular linters in other languages? I'm not sure about how general the context we should provide for JOSS is.

Author: There is a comprehensive summary of tools here: https://github.com/analysis-tools-dev/static-analysis

But I can't think of a good place where to place it, or how to cite it.

For example, expressions like `ifelse(x, TRUE, FALSE)` and `ifelse(x, FALSE, TRUE)` are redundant; just `x` or `!x` suffice in R code where logical vectors are a core data structure. The `redundant_ifelse_linter()` linter detects such discouraged usages.

```{r redundant_ifelse_linter}
lint(
  text = "ifelse(x >= 2.5, TRUE, FALSE)",
  linters = redundant_ifelse_linter()
)
```

- **Efficiency**

Sometimes the users might not be aware of a more efficient way offered by R for carrying out a computation. `{lintr}` offers linters to improve code efficiency by avoiding common inefficient patterns.

For example, the `any_is_na_linter()` linter detects usages of `any(is.na(x))` and suggests `anyNA(x)` as a more efficient alternative to detect presence of *any* missing values.

```{r any_is_na_linter_with_lint}
lint(
  text = "any(is.na(x), na.rm = TRUE)",
  linters = any_is_na_linter()
)
```

`anyNA()` in R is more efficient than `any(is.na())` because it stops execution once a missing value is found, while `is.na()` evaluates the entire vector.

```{r any_is_na_linter_without_lint}
lint(
  text = "anyNA(x)",
  linters = any_is_na_linter()
)

> ```

Reviewer: Below, we don't offer the "bad"/"good" pairing of snippets, should we remove this "good" example here?

Author: But this chunk is a companion to `any_is_na_linter_with_lint` chunk, which comes right before it.

Reviewer: Right... there's still some inconsistency. We have companion chunks for `comparison_negation_linter()` and `any_is_na_linter()`, but not `object_name_linter()` or `redundant_ifelse_linter()`. Since the corresponding "good" code for `redundant_ifelse_linter()` is covered in the text, maybe it's only `object_name_linter()` that's the outlier? Or do you think it's fine as is?

Author: Good point. I hadn't included the no-lint versions for these because they felt too trivial, but it does feel inconsistent to not see them. I have included them now.

- **Readability**

Coders spend significantly more time reading than writing code [@mcconnell2004code]. Thus, writing readable code makes the code more maintainable and reduces the possibility of introducing bugs stemming from a poor understanding of the code.

`{lintr}` provides a number of linters that suggest more readable alternatives. For example, `comparison_negation_linter()` blocks usages like `!(x == y)` where a direct relational operator is appropriate.

```{r comparison_negation_linter_with_lint}
lint(
  text = "!x == 2",
  linters = comparison_negation_linter()
)
```

> The more readable alternative here uses `!=`:

Reviewer: Blanking on the best phrasing here, but I think it's worth calling out that `!x==2` is even more unreadable than `!(x==2)`.

```suggestion
Note also the complicated operator precedence. The more readable alternative here uses `!=`:
```

```{r duplicate_argument_linter_with_lint}
lint(
  text = "mean(x = 1:5, x = 2:3)",
  linters = duplicate_argument_linter()
)
```

Even for cases where duplicate arguments are not an error, this linter explicitly discourages duplicate arguments.

```{r duplicate_argument_linter_without_lint}
lint(
  text = "list(x = TRUE, x = FALSE)",
  linters = duplicate_argument_linter()
)
```

This is because objects with duplicated names objects can be hard to work with programmatically and should typically be avoided.

```{r duplicate_arguments_example}
l <- list(x = TRUE, x = FALSE)
l["x"]
l[names(l) == "x"]
```

# Extensibility

`{lintr}` is designed for extensibility by allowing users to easily create custom linting rules.
There are two main ways to customize it:

  - Use additional arguments in existing linters. For example, although tidyverse style guide prefers snake_case for identifiers, if a project's conventions require it, the relevant linter can be customized to support it:

  ```{r object_name_linter_with_custom_style}
  lint(
    text = "my.var <- 1L",
    linters = object_name_linter(styles = "dotted.case")
  )
  ```

>   - Create new linters (by leveraging functions like `lintr::make_linter_from_xpath()`) tailored to match project-specific coding standards.

Reviewer: ```suggestion
  - Create new linters (by leveraging functions like `lintr::make_linter_from_xpath()`) tailored to match project- or organization-specific coding standards.
```
